File: 2024/Day02/Rust/Red-Nosed_Reports_P1.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader};



fn strs_to_int(strs: Vec<String>) -> Vec<i32> {
    let mut nums: Vec<i32> = Vec::new();
    for s in strs {
        nums.push(s.parse().unwrap());
    }
    return nums
}



fn main() -> io::Result<()> {
    let file = File::open("src/input.txt")?;
    let reader = BufReader::new(file);
    let mut safe_zones: Vec<String> = Vec::new();
    let mut unsafe_zones: Vec<String> = Vec::new();
    for line in reader.lines() {
        match line {
            Ok(line) => {
                let nums = strs_to_int(line.split_whitespace().map(String::from).collect());
                let mut arr = Vec::new();
                
               for i in 0..nums.len() - 1 {
                    if i == nums.len() - 1 {
                        break;
                    }
                    let diff = nums[i] - nums[i + 1];
                    if diff == 0 {
                        arr.push("so unsafe");
                    } else if diff > 0 && diff <= 3 {
                        arr.push("positive")
                    } else if diff < 0 && diff.abs() <= 3 {
                        arr.push("negative")
                    } else {
                        arr.push("too long");
                    }
                }
                if arr.iter().all(|x| arr[0] == *x) {
                    safe_zones.push(line.clone());
                } else {
                    unsafe_zones.push(line.clone());
                }
            },
            Err(e) => println!("Error: {}", e),
        }
    }
    println!("Safe zones: {}", safe_zones.len());
    println!("Unsafe zones: {}", unsafe_zones.len());
    Ok(())
    
}

File: 2024/Day02/Python/Red-Nosed_Reports_P1.py
file_name = "input_Red-Nosed_Reports_P1.txt"

try:
    with open(file_name, "r") as file:
        safe = 0
        while True:
            line = file.readline()
            if not line:
                break
            splited = list(map(int, line.split()))
            is_increase = True
            is_safe = True
            if (splited[0] > splited[1]):
                is_increase = False
            for i in range(1, len(splited)):
                if (is_increase):
                    if (splited[i] <= splited[i - 1] or splited[i] - splited[i - 1] > 3):
                        is_safe = False
                        break
                else:
                    if (splited[i] >= splited[i - 1] or splited[i - 1] - splited[i] > 3):
                        is_safe = False
                        break
            
            if (is_safe):
                safe += 1
        
        print(safe)
        

except:
    print("Error")